src/main/scala/io/qbeast/spark/delta/writer/SparkDeltaDataWriter.scala

```diff
@@ -0,0 +1,194 @@
+
+    val blockWriter =
+      BlockWriter(
+        dataPath = tableID.id,
+        schema = schema,
+        schemaIndex = qbeastData.schema,
+        factory = factory,
+        serConf = serConf,
+        qbeastColumns = qbeastColumns,
+        tableChanges = tableChanges)
+    qbeastData
+      .repartition(col(cubeColumnName))
+      .queryExecution
+      .executedPlan
+      .execute
+      .mapPartitions(blockWriter.writeRow)
+      .collect()
+      .toIndexedSeq
+
+  }
+
+  /**
+   * Split the files belonging to a cube into groups
+   * that lays between [MIN_FILE_SIZE_COMPACTION, MAX_FILE_SIZE_COMPACTION]
+   * Based on groupFilesIntoBins from Delta Lake OptimizeTableCommand
+   * @param cubeStatuses the cubes statuses present in the index
+   * @return
+   */
+  def groupFilesToCompact(
+      cubeStatuses: IISeq[(CubeId, IISeq[QbeastBlock])]): IISeq[(CubeId, IISeq[QbeastBlock])] = {
+
+    // Check what cubes are suitable for compaction
+    val cubesToCompact = cubeStatuses
+      .map(g => (g._1, g._2.filter(_.size >= MIN_FILE_SIZE_COMPACTION)))
+      .filter(_._2.nonEmpty)
+
+    cubesToCompact.flatMap { case (cube, blocks) =>
+      val groups = Seq.newBuilder[Seq[QbeastBlock]]
+      val g = Seq.newBuilder[QbeastBlock]
+      var count = 0L
+
+      blocks.foreach(b => {
+        if (b.size + count > MAX_FILE_SIZE_COMPACTION) {
+          // If we reach the MAX_FILE_SIZE_COMPACTION limit
+          // we output a group of files for that cube
+          groups += g.result()
+          // Clear the current group
+          g.clear()
+          // Clear the count
+          count = 0L
+        }
+        // Add the block to the group
+        // Sum the size of the block
+        g += b
+        count += b.size
+      })
+      groups += g.result() // Add the last group
+      groups.result().map(b => (cube, b.toIndexedSeq))
+    }
+  }
+
+  /**
+   * Compact the files.
+   * Method based on Delta Lake OptimizeTableCommand
+   * Experimental: the implementation may change by using the CubeID as partition key
+   * and delegating the further execution to the underlying Format
+   *
+   * @param tableID
+   * @param schema
+   * @param data
+   * @param tableChanges
+   * @return
+   */
+  override def compact(
+      tableID: QTableID,
+      schema: StructType,
+      indexStatus: IndexStatus,
+      tableChanges: TableChanges): IISeq[FileAction] = {
+
+    val sparkSession = SparkSession.active
+
+    val cubesToCompact = indexStatus.cubesStatuses.mapValues(_.files).toIndexedSeq
+    val cubesToCompactGrouped = groupFilesToCompact(cubesToCompact)
+
+    val parallelJobCollection = new ParVector(cubesToCompactGrouped.toVector)
+
+    // Create a thread pool to be able to parallelize the compaction task
+    // Each thread will be in charge of compacting the set of files for one single cube
+    val threadPool =
+      QbeastThreadUtils.threadUtils.newForkJoinPool("Compaction", maxThreadNumber = 15)
```

Comment:
Normally functions should not create thread pools for every call. Thread in JVM is an expensive abstraction, so the less threads we have the better. The application should control the maximum number of the threads. Here in theory there can be multiplicative effect when the method is called in parallel by several clients. I would prefer to have one dedicated thread pool with some configurable number of threads, the rule of thumb is 2 threads per core.

This part should be improved in the future, possibly as a separate issue in order not to block this PR.